authz/audit_logger.go

```diff
@@ -0,0 +1,90 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package authz
+
+import (
+	"sync"
+)
+
+var (
+	mu sync.Mutex
+	m  = make(map[string]AuditLoggerBuilder)
+)
+
+// RegisterAuditLoggerBuilder registers the builder in a global map
+// using b.Name() as the key.
+// This should only be called during initialization time (i.e. in an init() function).
+// If multiple builders are registered with the same name, the one registered last
+// will take effect.
+func RegisterAuditLoggerBuilder(b AuditLoggerBuilder) {
+	mu.Lock()
+	defer mu.Unlock()
+	m[b.Name()] = b
+}
+
+// AuditInfo contains information used by the audit logger during an audit logging event.
+type AuditInfo struct {
+	// RPCMethod is the method of the audited RPC.
+	RPCMethod string `json:"rpc_method,omitempty"`
+	// Principal is the identity of the RPC. Currently it will only be available in
+	// certificate-based TLS authentication.
+	Principal string `json:"principal,omitempty"`
+	// PolicyName is the authorization policy name (or the xDS RBAC filter name).
+	PolicyName string `json:"policy_name,omitempty"`
+	// MatchedRule is the matched rule (or policy name in the xDS RBAC filter). It will be
+	// empty if there is no match.
+	MatchedRule string `json:"matched_rule,omitempty"`
+	// Authorized indicates whether the audited RPC is authorized or not.
+	Authorized bool `json:"bool,omitempty"`
+}
+
+// AuditLoggerConfig defines the configuration for a particular implementation of audit logger.
+type AuditLoggerConfig interface {
+	// Name returns the same name as that returned by its supported builder.
+	Name() string
+}
+
+// AuditLogger is the interface for an audit logger.
+type AuditLogger interface {
+	// Log logs the auditing event with the given information.
+	Log(*AuditInfo)
```

Comment:
If this is supposed to be allowed to block, I believe we do want a context here.  Otherwise, there is no way to cancel it or let it know what deadline it has.

Reply:
Thanks for the review, Doug! I will work on the other comments but I'd like to discuss this particular comment a bit more. So `Log()` will be called during the authorization process, and I think we agreed that the same function in C++ and Java will only take the equivalent of this `AuditInfo`. So I believe there is no way there as well for the function to cancel or know anything about the deadline? If this is true, I feel we should keep it consistent here, although I do like passing a context to enable cancellation, which was why I had included it.

Reply:
This seems like this is going to be a language difference to me, as the threading models are different in the different languages.  All blocking functions in Go should ideally accept contexts, but no similar constraint exists in Java/C++.

Reply:
This makes sense. I noticed methods like `IsAuthorized()` in this pkg also takes a context whereas I don't believe C++/Java has something similar. I added the context back.